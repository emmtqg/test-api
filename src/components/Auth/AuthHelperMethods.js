import decode from 'jwt-decode';

import config from '../Config';

// The apiCamelizedLookup provides a fieldname check and crossreference
// for the form field names. It aids in translation of valid property
// names to whatever the API required (single point of definition between
// the state properties and api names). It was assumed if the ESLint rules
// did not allow underscores (as specified in the API names) then, rather
// than overriding the linter, a mechanism should be incorporated to allow
// any API to specify properties as required and they can then be updated
// in one location. If form fields are checked for testing, then the form
// component can also utilize the apiCamelizedLookup for field specifications.
const { apiCamelizedLookup } = config;
const apiNames = Object.keys(apiCamelizedLookup);

export default class AuthHelperMethods {
  getApiPropName = (name) => {
    // Get the API 'expected' property of input form field
    let index = -1;
    Object.values(apiCamelizedLookup).findIndex((ele, i) => {
      index = i;
      return (ele === name);
    });
    return index; // returning matched index or -1 if not in apiCamelized object
  }

  lookupApiName = (name) => {
    const index = this.getApiPropName(name);
    if (index > -1) {
      // apiNames are the API keys in lookup array
      return apiNames[index];
    }

    return ''; // API will error with invalid fieldname
  }

  register = (registerObject) => {
    const {
      firstName,
      lastName,
      email,
      password,
      confirmPassword,
    } = registerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.registerUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          email,
          password,
          [this.lookupApiName('confirmPassword')]: confirmPassword,
        }),
      })
        .then((response) => {
          // console.log(`Auth (Register): ${response}`);
          if (response) {
            this.setToken(response.token); // Setting the token in localStorage
            return Promise.resolve(response);
          }
          return Promise.reject(response);
        })
    );
  }

  newPlayer = (newPlayerObject) => {
    const {
      firstName,
      lastName,
      rating,
      handedness,
    } = newPlayerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.playersUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          rating,
          handedness,
        }),
      })
    );
  }

  getPlayers = () => (
    this.fetch(config.playersUrl, {
      method: 'GET',
    })
  );

  removePlayer = (playerId) => {
    const deleteUrl = `${config.playersUrl}/${playerId}`;
    return (
      this.fetch(deleteUrl, {
        method: 'DELETE',
      })
    );
  }

  login = (loginObject) => {
    const {
      email,
      password,
    } = loginObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.loginUrl, {
        method: 'POST',
        body: JSON.stringify({
          email,
          password,
        }),
      })
        .then((response) => {
          if (response) {
            // console.log(`Auth (Login): ${response}`);
            this.setToken(response.token); // Setting the token in localStorage
            return response;
          }
          return Promise.reject(response);
        })
        .catch((error) => {
          if (error && error.message) {
            // console.log(`Error: ${error.message}`);
          }
          return Promise.reject(error);
        })
    );
  }

  loggedIn = () => {
    // Checks if there is a saved token and it's still valid
    const token = this.getToken(); // Getting token from localstorage
    return (typeof token !== typeof undefined) && (!!token && !this.isTokenExpired(token));
  }

  isTokenExpired = (token) => {
    try {
      const decoded = decode(token);
      if (decoded.exp < Date.now() / 1000) { // Checking if token is expired.
        return true;
      }
      return false;
    } catch (err) {
      // console.log('Check for expired tolen failed.');
      return false;
    }
  }

  setToken = (idToken) => {
    // Saves user token to localStorage
    localStorage.setItem('id_token', idToken);
  }

  getToken = () =>
    // Retrieves the user token from localStorage
    localStorage.getItem('id_token')


  logout = () => {
    // Clear user token and profile data from localStorage
    localStorage.removeItem('id_token');
  }

  getConfirm = () => {
    // Using jwt-decode npm package to decode the token
    const answer = decode(this.getToken());
    // console.log(`Get confirm token => ${answer}`);
    return answer;
  }

  fetch = (url, options) => {
    // performs api calls sending the required authentication headers
    const headers = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    };

    // Setting Authorization header
    // Authorization: Bearer xxxxxxx.xxxxxxxx.xxxxxx
    if (this.loggedIn()) {
      headers.Authorization = `Bearer ${this.getToken()}`;
    }

    return fetch(url, {
      headers,
      ...options,
    })
      .then(this.checkStatus)
      .then(response => response.json())
      .catch((error) => {
        // console.log(`Error: ${error.message}`);
        throw error;
      });
  }

  checkStatus = (response) => {
    // Success status lies between 200 to 300
    if (response.status >= 200 && response.status < 300) {
      return response;
    }

    // raises an error in case response status is not a success
    const error = new Error(response.statusText);
    error.response = response;
    error.status = response.status;
    throw error;
  }
}
