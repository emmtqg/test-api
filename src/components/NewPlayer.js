/* eslint-disable no-alert */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormHelperText from '@material-ui/core/FormHelperText';
import Select from '@material-ui/core/Select';
import Add from '@material-ui/icons/Add';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import withStyles from '@material-ui/core/styles/withStyles';
import green from '@material-ui/core/colors/green';

import './SideBySideComponents.css';
import config from './Config';

const { apiCamelizedLookup } = config;
const styles = theme => ({
  main: {
    width: 'auto',
    display: 'block', // Fix IE 11 issue.
    // marginLeft: theme.spacing.unit * 3,
    // marginRight: theme.spacing.unit * 3,
    [theme.breakpoints.up((400 + theme.spacing.unit) * 3 * 2)]: {
      width: 400,
      marginLeft: 'auto',
      marginRight: 'auto',
    },
  },
  paper: {
    marginTop: theme.spacing.unit * 8,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 3}px ${theme.spacing.unit * 3}px`,
  },
  avatar: {
    margin: theme.spacing.unit,
    backgroundColor: green[400],
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing.unit,
  },
  submit: {
    marginTop: theme.spacing.unit * 3,
  },
  selectEmpty: {
    marginTop: theme.spacing.unit * 2,
  },
});

// Values get passed to the API
const handedness = {
  NONE_SPECIFIED: 'none',
  LEFT: 'left',
  RIGHT: 'right',
};

class NewPlayer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      firstName: '',
      lastName: '',
      rating: 0,
      handedness: handedness.NONE_SPECIFIED,
      remember: false,
      isAuthorized: false,
      isLoading: false,
      message: '',
      messageSuggest: '',
      labelWidth: 0,
    };

    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  componentWillMount() {
    // Here is a great place to redirect someone who is already logged in
    if (!this.props.auth.loggedIn()) {
      this.props.history.replace('/');
    }
  }

  handleChange = (event) => {
    // 'normalize' the underscore id's with the
    // state property fields to both satify project
    // API requirements and ESLINT rules.
    const { name } = event.target;
    const stateName = apiCamelizedLookup[name];
    this.setState({
      [stateName]: event.target.value,
    });
  }

  handleCancel = (e) => {
    e.preventDefault();
    this.props.history.replace('/roster');
  }

  handleFormSubmit = (e) => {
    e.preventDefault();

    if (!(this.state.handedness in handedness) &&
         (this.state.handedness === handedness.NONE_SPECIFIED)) {
      this.setState({
        message: 'Please make a selection for the perferred paddle hand.',
      });

      return null;
    }

    this.props.auth.newPlayer(this.state)
      .then((response) => {
        if (typeof response === typeof undefined) {
          this.setState({
            message: 'Could not create the player.',
            isLoading: false,
          });
        }
        const { player } = response;
        this.setState({
          message: `New team member ${player.firstName} ${player.lastName} is now on the team!`,
          messageSuggestion: 'Don\'t forget to get the a t-shirt!',
          isLoading: false,
        });

        return this.props.history.replace('/roster');
      })
      .catch((error) => {
        let message = `Adding a new player encountered an error: (${error.status}) ${error.message}.`;
        if (error.status === 409) {
          message = `This player already exsists in our player data. (${error.status}) ${error.message} was returned from the server.`;
        }
        this.setState({
          message,
          messageSuggest: 'Please recheck your player information.',
        });
      });
    return null;
  }

  render() {
    const { classes } = this.props;

    return (
      <main className={classes.main}>
        <CssBaseline />
        <Paper className={classes.paper}>
          <Avatar className={classes.avatar}>
            <Add />
          </Avatar>
          <Typography component="h1" variant="h5">
            New Player
          </Typography>
          <Typography style={{ color: 'red' }} component="h6" variant="h6">
            {this.state.message}<br />
            {this.state.messageSuggest}
          </Typography>
          <form className={classes.form} onSubmit={this.handleFormSubmit}>
            <div className="flexGrid">
              <div className="col">
                <FormControl margin="normal" required className="almostWidth">
                  <InputLabel htmlFor="first_name">First Name</InputLabel>
                  <Input id="first_name" name="first_name" autoComplete="first_name" autoFocus onChange={this.handleChange} />
                </FormControl>
              </div>
              <div className="col">
                <FormControl margin="normal" required className="justifyEnd">
                  <InputLabel htmlFor="last_name">Last Name</InputLabel>
                  <Input id="last_name" name="last_name" autoComplete="last_name" onChange={this.handleChange} />
                </FormControl>
              </div>
            </div>
            <FormControl margin="normal" required fullWidth>
              <InputLabel htmlFor="email">Rating</InputLabel>
              <Input id="rating" name="rating" autoComplete="rating" onChange={this.handleChange} />
            </FormControl>
            <FormControl className={classes.formControl}>
              <InputLabel htmlFor="handedness">Choose</InputLabel>
              <Select
                value={this.state.handedness}
                onChange={this.handleChange}
                required
                fullWidth
                id="handedness"
                inputProps={{
                  name: 'handedness',
                  id: 'handedness',
                }}
              >
                <MenuItem value="" id="handedness.NONE">
                  <em>None</em>
                </MenuItem>
                <MenuItem value={handedness.RIGHT}>Right</MenuItem>
                <MenuItem value={handedness.LEFT}>Left</MenuItem>
              </Select>
              <FormHelperText>Preferred Paddle Hand</FormHelperText>
            </FormControl>
            <Button
              type="submit"
              fullWidth
              variant="contained"
              color="primary"
              className={classes.submit}
              onSubmit={this.handleSubmit}
            >
              Add Player
            </Button>
            <br /><br />
            <Button
              fullWidth
              variant="contained"
              color="secondary"
              onClick={this.handleCancel}
            >
              Cancel
            </Button>
          </form>
        </Paper>
      </main>
    );
  }
}

NewPlayer.defaultProps = {
  auth: {},
  // History object passed via Router or a function
  // that passed in unit tests via jest.
  history: PropTypes.oneOfType([
    PropTypes.shape({
      replace: () => {},
    }),
    () => {},
  ]),
};

NewPlayer.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  classes: PropTypes.object.isRequired,
  auth: PropTypes.shape({
    loggedIn: PropTypes.func.isRequired,
    newPlayer: PropTypes.func.isRequired,
  }),
  // History object passed via Router or a function
  // that passed in unit tests via jest.
  history: PropTypes.oneOfType([
    PropTypes.shape({
      replace: PropTypes.func,
    }),
    PropTypes.func,
  ]),
};

export default withStyles(styles)(NewPlayer);
