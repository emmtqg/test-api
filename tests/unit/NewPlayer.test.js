import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import NewPlayer from '../../src/components/NewPlayer';
import AuthHelperMethods from '../../src/components/Auth/AuthHelperMethods';

// All field forms are translated between the form and API property
// names and the component state properties. This is defined in the
// project Config.js file.
const apiCamelizedLookupForNewPlayer = {
  first_name: 'firstName',
  last_name: 'lastName',
  handedness: 'handedness',
  rating: 'rating',
};

let UnwrappedShallowComponent;
let wrapper;

// auth mocks
const newPlayerMock = jest.fn();
newPlayerMock.mockResolvedValue({
  success: true,
  token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI1Y2E3NzI0YTMxMGQyMTdmZjM5MDFlYWUiLCJpYXQiOjE1NTQ0Nzc2NDJ9.dB--qqJQT6Zx8Ccn6HhKAXqtZX0oqhgyV2oEjy3RonU',
  user: {
    first_name: 'Same',
    last_name: 'Smith',
    email: 'a@b.com',
    id: '5ca7724a310d217ff3901eae',
  },
});
// expect(mock('foo')).resolves.toBe('bar');

// history mock
const history = {
  replace: jest.fn(),
};
const auth = new AuthHelperMethods();
auth.loggedIn = jest.fn(() => true);

beforeAll(() => {
  auth.newPlayer = jest.fn(() => { Promise.resolve({}); });
  wrapper = mount(<NewPlayer auth={auth} props={history} />);
  const shallowComponent = shallow(<NewPlayer auth={auth} props={history} />);
  UnwrappedShallowComponent = shallowComponent.first().shallow();
});

afterAll(() => {
  wrapper.unmount();
});


describe('The New Player Form should', () => {
  it('match the New Player form snapshot', () => {
    expect(wrapper.find('NewPlayer').length).toEqual(1);
  });

  it('render correctly', () => {
    const component = renderer.create(<NewPlayer auth={auth} props={history} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('actually have a form element', () => {
    expect(wrapper.find('form').length).toBe(1);
  });

  // Loop through the apiCamelizedLookup array for all field forms
  describe('The form fields retrieve data and update component state', () => {
    // state names are transformed to camelCase for eslint rule
    // This is done via an enum like object defined in the project
    // Config.js file (under /src).
    const formNames = Object.keys(apiCamelizedLookupForNewPlayer);
    const formValues = Object.values(apiCamelizedLookupForNewPlayer);
    let input;
    formNames.forEach((element) => {
      it(` checking ${element} has a defined onChange function`, () => {
        // MUI returning multiple nodes with #text_field, get the first with props
        input = wrapper.find(`#${element}`).first();
        expect(input.length).toBeGreaterThanOrEqual(1);
        expect(input.props('onChange')).toBeDefined();
        expect(typeof input.props('onChange').onChange === 'function').toBe(true);
      });
    });

    formNames.forEach((element, index) => {
      it(` checking ${element} updates component state on change`, () => {
        input = UnwrappedShallowComponent.find(`#${element}`).first();
        if ((element === 'first_name') || (element === 'last_name')) {
          input.simulate('change', { target: { name: element, value: 'Ping' } });
          expect(UnwrappedShallowComponent.state(formValues[index])).toBe('Ping');
        } else if (element === 'rating') {
          input.simulate('change', { target: { name: element, value: '9000' } });
          expect(UnwrappedShallowComponent.state(formValues[index])).toBe('9000');
        } else if (element === 'handedness') {
          input.simulate('change', { target: { name: element, value: 'right' } });
          expect(UnwrappedShallowComponent.state(formValues[index])).toBe('right');
        } else {
          // flags erroneous field listed in formNames (apiCamelizedLookupForNewPlayer)
          expect(true).toBe(false);
        }
      });
    });
  });
});
