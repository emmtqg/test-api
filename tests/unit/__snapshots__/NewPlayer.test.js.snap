// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The New Player Form should render correctly 1`] = `
<main
  className="NewPlayer-main-1"
>
  <div
    className="MuiPaper-root-7 MuiPaper-elevation2-11 MuiPaper-rounded-8 NewPlayer-paper-2"
  >
    <div
      className="MuiAvatar-root-34 MuiAvatar-colorDefault-35 NewPlayer-avatar-3"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-37"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
    <h1
      className="MuiTypography-root-46 MuiTypography-h5-62"
    >
      New Player
    </h1>
    <h6
      className="MuiTypography-root-46 MuiTypography-h6-63"
      style={
        Object {
          "color": "red",
        }
      }
    >
      
      <br />
      
    </h6>
    <form
      className="NewPlayer-form-4"
      onSubmit={[Function]}
    >
      <div
        className="flexGrid"
      >
        <div
          className="col"
        >
          <div
            className="MuiFormControl-root-82 MuiFormControl-marginNormal-83 almostWidth"
          >
            <label
              className="MuiFormLabel-root-97 MuiFormLabel-required-102 MuiInputLabel-required-90 MuiInputLabel-root-86 MuiInputLabel-formControl-91 MuiInputLabel-animated-94"
              data-shrink={false}
              htmlFor="first_name"
            >
              First Name
              <span
                className="MuiFormLabel-asterisk-103"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-117 MuiInput-root-104 MuiInput-underline-108 MuiInputBase-formControl-118 MuiInput-formControl-105"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="first_name"
                autoFocus={true}
                className="MuiInputBase-input-127 MuiInput-input-112"
                disabled={false}
                id="first_name"
                name="first_name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          className="col"
        >
          <div
            className="MuiFormControl-root-82 MuiFormControl-marginNormal-83 justifyEnd"
          >
            <label
              className="MuiFormLabel-root-97 MuiFormLabel-required-102 MuiInputLabel-required-90 MuiInputLabel-root-86 MuiInputLabel-formControl-91 MuiInputLabel-animated-94"
              data-shrink={false}
              htmlFor="last_name"
            >
              Last Name
              <span
                className="MuiFormLabel-asterisk-103"
              >
                 *
              </span>
            </label>
            <div
              className="MuiInputBase-root-117 MuiInput-root-104 MuiInput-underline-108 MuiInputBase-formControl-118 MuiInput-formControl-105"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="last_name"
                className="MuiInputBase-input-127 MuiInput-input-112"
                disabled={false}
                id="last_name"
                name="last_name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiFormControl-root-82 MuiFormControl-marginNormal-83 MuiFormControl-fullWidth-85"
      >
        <label
          className="MuiFormLabel-root-97 MuiFormLabel-required-102 MuiInputLabel-required-90 MuiInputLabel-root-86 MuiInputLabel-formControl-91 MuiInputLabel-animated-94"
          data-shrink={false}
          htmlFor="email"
        >
          Rating
          <span
            className="MuiFormLabel-asterisk-103"
          >
             *
          </span>
        </label>
        <div
          className="MuiInputBase-root-117 MuiInput-root-104 MuiInput-underline-108 MuiInputBase-formControl-118 MuiInput-formControl-105"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="rating"
            className="MuiInputBase-input-127 MuiInput-input-112"
            disabled={false}
            id="rating"
            name="rating"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-82"
      >
        <label
          className="MuiFormLabel-root-97 MuiFormLabel-filled-101 MuiInputLabel-root-86 MuiInputLabel-formControl-91 MuiInputLabel-animated-94 MuiInputLabel-shrink-93"
          data-shrink={true}
          htmlFor="handedness"
        >
          Choose
        </label>
        <div
          className="MuiInputBase-root-117 MuiInput-root-104 MuiInput-underline-108 MuiInputBase-fullWidth-126 MuiInput-fullWidth-111 MuiInputBase-formControl-118 MuiInput-formControl-105"
          onClick={[Function]}
          required={true}
        >
          <div
            className="MuiSelect-root-134"
          >
            <div
              aria-haspopup="true"
              aria-pressed="false"
              className="MuiSelect-select-135 MuiSelect-selectMenu-138 MuiInputBase-input-127 MuiInput-input-112"
              id="select-handedness"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              id="handedness"
              name="handedness"
              type="hidden"
              value="none"
            />
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-37 MuiSelect-icon-140"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <p
          className="MuiFormHelperText-root-145 MuiFormHelperText-filled-151"
        >
          Preferred Paddle Hand
        </p>
      </div>
      <button
        className="MuiButtonBase-root-179 MuiButton-root-153 MuiButton-contained-164 MuiButton-containedPrimary-165 MuiButton-raised-167 MuiButton-raisedPrimary-168 MuiButton-fullWidth-178 NewPlayer-submit-5"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="submit"
      >
        <span
          className="MuiButton-label-154"
        >
          Add Player
        </span>
        <span
          className="MuiTouchRipple-root-182"
        />
      </button>
      <br />
      <br />
      <button
        className="MuiButtonBase-root-179 MuiButton-root-153 MuiButton-contained-164 MuiButton-containedSecondary-166 MuiButton-raised-167 MuiButton-raisedSecondary-169 MuiButton-fullWidth-178"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-154"
        >
          Cancel
        </span>
        <span
          className="MuiTouchRipple-root-182"
        />
      </button>
    </form>
  </div>
</main>
`;
